language/vm/vm_runtime/vm_runtime_types/src/value.rs

```diff
@@ -475,93 +661,275 @@ impl GlobalRef {
+    /// locals vector.
+    pub fn new(size: usize) -> Self {
+        Locals(vec![ValueImpl::Invalid; size])
+    }
+
+    /// Copy the `Value` at index `idx`. Return an error if index is out of bounds.
+    pub fn copy_loc(&self, idx: usize) -> VMResult<Value> {
+        if let Some(local_ref) = self.0.get(idx) {
+            local_ref.copy_value()
+        } else {
+            let msg = format!(
+                "Index {} out of bounds for {} while indexing {}",
+                idx,
+                self.0.len(),
+                IndexKind::LocalPool
+            );
+            Err(VMStatus::new(StatusCode::INDEX_OUT_OF_BOUNDS).with_message(msg))
+        }
+    }
 
-impl From<MutVal> for Option<u64> {
-    fn from(value: MutVal) -> Option<u64> {
-        match &*value.peek() {
-            Value::U64(i) => Some(*i),
-            _ => None,
+    /// Move the `Value` at index `idx` out of the locals vector. The local is replaced with
+    /// `ValueImpl::Invalid`. Return an error if index is out of bounds.
+    pub fn move_loc(&mut self, idx: usize) -> VMResult<Value> {
+        if let Some(local_ref) = self.0.get_mut(idx) {
+            let old_local = replace(local_ref, ValueImpl::Invalid);
+            old_local.into_value()
+        } else {
+            let msg = format!(
+                "Index {} out of bounds for {} while indexing {}",
+                idx,
+                self.0.len(),
+                IndexKind::LocalPool
+            );
+            Err(VMStatus::new(StatusCode::INDEX_OUT_OF_BOUNDS).with_message(msg))
         }
     }
-}
 
-impl From<MutVal> for Option<bool> {
-    fn from(value: MutVal) -> Option<bool> {
-        match &*value.peek() {
-            Value::Bool(b) => Some(*b),
-            _ => None,
+    /// Store `value` in input in the local at index `idx`.
+    /// Return an error if index is out of bounds.
+    pub fn store_loc(&mut self, idx: usize, value: Value) -> VMResult<()> {
+        if let Some(local_ref) = self.0.get_mut(idx) {
+            replace(local_ref, value.0);
+            Ok(())
+        } else {
+            let msg = format!(
+                "Index {} out of bounds for {} while indexing {}",
+                idx,
+                self.0.len(),
+                IndexKind::LocalPool
+            );
+            Err(VMStatus::new(StatusCode::INDEX_OUT_OF_BOUNDS).with_message(msg))
         }
     }
-}
 
-impl From<MutVal> for Option<AccountAddress> {
-    fn from(value: MutVal) -> Option<AccountAddress> {
-        match *value.peek() {
-            Value::Address(addr) => Some(addr),
-            _ => None,
+    /// Borrow the local at index `idx`.
+    /// Return an error if index is out of bounds.
+    ///
+    /// Current implementation "promotes" a local value to a reference "on the fly".
+    /// First call to borrow_global converts the value into a reference (`ValueImpl::LocalRef`).
+    pub fn borrow_loc(&mut self, idx: usize) -> VMResult<Value> {
```

Comment:
Why is it taking a `&mut self` rather than `&self`, IMHO this function won't mutate the local vector.

Reply:
its calling a get_mut on the locals to promote the value to a reference so I thin it's needed, no?